utility/SPIFirmata.cpp

```diff
@@ -0,0 +1,211 @@
+  modify it under the terms of the GNU Lesser General Public
+  License as published by the Free Software Foundation; either
+  version 2.1 of the License, or (at your option) any later version.
+
+  See file LICENSE.txt for further informations on licensing terms.
+
+  Last updated January 15th, 2017
+*/
+
+#include "SPIFirmata.h"
+
+SPIFirmata::SPIFirmata()
+{
+  init();
+}
+
+void SPIFirmata::init()
+{
+  mDeviceId = 0;
+  mCsPin = -1;
+  mCsActiveState = SPI_CS_ACTIVE_LOW; // default
+}
+
+bool SPIFirmata::handlePinMode(uint8_t pin, int mode)
+{
+  // There is no reason for a user to manually set the SPI pin modes
+  return false;
+}
+
+void SPIFirmata::handleCapability(uint8_t pin)
+{
+  // ignore SS pin for now
+  if (IS_PIN_SPI(pin) && pin != SS) {
+    Firmata.write(PIN_MODE_SPI);
+    // would actually use a value that corresponds to a specific pin (MOSI, MISO, SCK)
+    // for now, just set to 1
+    Firmata.write(1);
+  }
+}
+
+bool SPIFirmata::handleSysex(uint8_t command, uint8_t argc, uint8_t *argv)
+{
+  if (command == SPI_DATA) {
+    uint8_t mode = argv[0];
+    // not using channel yet
+    uint8_t channel = argv[1] & SPI_CHANNEL_MASK;
+
+    switch (mode) {
+      case SPI_CONFIG:
+        SPI.begin();
+        // SPI pin states are configured by SPI.begin, but we still register them with Firmata.
+        Firmata.setPinMode(MOSI, PIN_MODE_SPI);
+        Firmata.setPinMode(MISO, PIN_MODE_SPI);
+        Firmata.setPinMode(SCK, PIN_MODE_SPI);
+        // ignore SS for now
+        //Firmata.setPinMode(SS, PIN_MODE_SPI);
+        break;
+      case SPI_BEGIN_TRANSACTION:
+      {
+        mDeviceId = argv[1] >> 2;
+        uint8_t bitOrder = argv[2] & SPI_BIT_ORDER_MASK;
+        uint8_t dataMode = argv[2] >> 1;
+        uint32_t clockSpeed = (uint32_t)argv[3] | ((uint32_t)argv[4] << 7) |
+            ((uint32_t)argv[5] << 14) | ((uint32_t)argv[6] << 21) | ((uint32_t)argv[7] << 28);
+
+        // argv[8] = wordSize, but not currently used since SPI.transfer only uses 8-bit words
+
+        if (argc > 9) {
+          mCsPin = argv[9];
+          pinMode(mCsPin, OUTPUT);
+
+          if (argv[10] != END_SYSEX) {
+            mCsActiveState = argv[10];
+          } else {
+            // set default
+            mCsActiveState = SPI_CS_ACTIVE_LOW;
+          }
+          // set CS pin to opposite of active state
+          digitalWrite(mCsPin, !mCsActiveState);
+
+          // protect the CS pin
+          // TODO - decide if this is the best approach. If PIN_MODE_SPI is set, the user cannot
+          // manually control the CS pin using DIGITAL_MESSAGE.
+          Firmata.setPinMode(mCsPin, PIN_MODE_SPI);
+        }
+
+        SPISettings settings(clockSpeed, getBitOrder(bitOrder), getDataMode(dataMode));
+        SPI.beginTransaction(settings);
+        break;
+      }
+      case SPI_END_TRANSACTION:
+        SPI.endTransaction();
+        break;
+      case SPI_TRANSFER:
```

Comment:
TBD whether or not to keep this a single command or to break it up into 3 separate commands:

```
SPI_READ
SPI_WRITE
SPI_TRANSFER (normal SPI operation - read a word for every word written)
```

Another option would be to specify both an RX buffer size and a TX buffer size instead of specifying read/write modes. If only the RX buffer size is specified, then write 0x00 for each word to be read. If only the TX buffer size is specified, then write but ignore the return values. If TX buffer = RX buffer, replace the write word with the reply word and return the buffer. If RX buffer size is > TX buffer size, write TX words first, ignoring reply, then write 0x00 for each RX word. TBD how if there is a use case for RX buffer size < TX buffer size.